File: using_mlir.py
# Load Reactant.jl-generated StableHLO/MLIR and use it in Python

from common import load_mlir, load_mlir_enzyme  # caps GPU memory before importing jax

import numpy as np
import jax
import jax.numpy as jnp


x = np.array([1.0, 2.0, 3.0], dtype=np.float32)

# direct jax, run only:
loss = load_mlir("negsumsq_reactant.mlir")
loss(x)

# enzyme-ad, can run and autodiff
loss_e = load_mlir_enzyme("negsumsq_reactant.mlir")
scalar = lambda v: loss_e(v)[0]
xj = jnp.asarray(x)
float(jax.jit(scalar)(xj))
jax.jit(jax.grad(scalar))(xj)

File: using_mlir.jl
# Load JAX-generated StableHLO/MLIR and use it in Julia

include("common.jl")            # caps GPU memory before the first `using Reactant`; defines load_mlir

using Reactant
using Reactant: to_rarray
using Enzyme

f_loss = load_mlir("negsumsq_jax.mlir")

x = Float32[1, 2, 3]
rx = to_rarray(x)

# Note the difference between the loaded MLIR code and the Reactant-optimized
# MLIR code:
@code_hlo optimize=false f_loss(rx)
@code_hlo f_loss(rx)

loss = @jit f_loss(rx)

# We can still differentiate loaded MLIR with Enzyme:
@code_hlo Enzyme.gradient(Enzyme.Reverse, f_loss, rx)
∇loss = @jit Enzyme.gradient(Enzyme.Reverse, f_loss, rx)
